work/v1/types.go

```diff
@@ -225,3 +225,68 @@ type ManifestWorkList struct {
 	// Items is a list of manifestworks.
 	Items []ManifestWork `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:subresource:status
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AppliedManifestWork represents an applied manifestwork on managed cluster. It is placed
+// on managed cluster. An AppliedManifestWork links to a manifestwork on a hub recording resources
+// deployed in the managed cluster.
+// When the agent is removed from managed cluster, cluster-admin on managed cluster
+// can delete appliedmanifestwork to remove resources deployed by the agent.
+type AppliedManifestWork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec represents the desired configuration of AppliedManifestWork
+	Spec AppliedManifestWorkSpec `json:"spec,omitempty"`
+
+	// Status represents the current status of AppliedManifestWork
+	// +optional
+	Status AppliedManifestWorkStatus `json:"status,omitempty"`
+}
+
+// AppliedManifestWorkSpec represents the desired configuration of AppliedManifestWork
+type AppliedManifestWorkSpec struct {
+	// HubID represents the ID of the hub that the agent fetches ManifestWork. It could
+	// be a hash of hub kubernetes apiserver url.
+	// +required
+	HubID string `json:"hubID"`
+
+	// ManifestWorkName represents the name of the related manifestwork on hub.
+	// +required
+	ManifestWorkName string `json:"manifestWorkName"`
+
+	// ClusterName represents the name of the cluster on hub where the related manifestwork
+	// is placed.
+	// +required
+	ClusterName string `json:"clusterNamee"`
+}
+
+// AppliedManifestWorkStatus represents the current status of AppliedManifestWork
+type AppliedManifestWorkStatus struct {
+	// AppliedResources represents a list of resources defined within the manifestwork that are applied.
+	// Only resources with valid GroupVersionResource, namespace, and name are suitable.
+	// An item in this slice is deleted when there is no mapped manifest in manifestwork.Spec or by finalizer.
+	// The resource relating to the item will also be removed from managed cluster.
+	// The deleted resource may still be present until the finalizers for that resource are finished.
+	// However, the resource will not be undeleted, so it can be removed from this list and eventual consistency is preserved.
+	// +optional
+	AppliedResources []AppliedManifestResourceMeta `json:"appliedResources,omitempty"`
```

Comment:
After we introduce `AppliedManifestWork`, applied resources of ManifestWork are stored in two different places: `ManifestWork.Status.AppliedResrouces` and `AppliedManifestWork.Status.AppliedResrouces`. Changes of this slice may come from `ManifestWork` when applying manifests, or from `AppliedManifestWork` when deleting applied resources. It's complicated to sync them with each other. So I suggest we only keep one copy of this slice in `AppliedManifestWork.Status`.

Reply:
That's a good point and I'm still thinking about whether we need to keep the information in a single place, or whether we should clearly define the precedence of AppliedManifestWork and ManifestWork.Status

Reply:
I think this is a good point.  Since we only act on the information on the spoke, I think it's a good point.  It does mean that the update is non-atomic, but I think that's ok.